v4/supervisor.go

```diff
@@ -0,0 +1,956 @@
+package suture
+
+// FIXMES in progress:
+// 1. Ensure the supervisor actually gets to the terminated state for the
+//     unstopped service report.
+// 2. Save the unstopped service report in the supervisor.
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"math"
+	"math/rand"
+	"runtime"
+	"sync"
+	"time"
+)
+
+const (
+	notRunning = iota
+	normal
+	paused
+	terminated
+)
+
+type supervisorID uint32
+type serviceID uint32
+
+/*
+Supervisor is the core type of the module that represents a Supervisor.
+
+Supervisors should be constructed either by New or NewSimple.
+
+Once constructed, a Supervisor should be started in one of three ways:
+
+ 1. Calling .Serve(ctx).
+ 2. Calling .ServeBackground(ctx).
+ 3. Adding it to an existing Supervisor.
+
+Calling Serve will cause the supervisor to run until the passed-in
+context is cancelled. Often one of the last lines of the "main" func for a
+program will be to call one of the Serve methods.
+
+Calling ServeBackground will CORRECTLY start the supervisor running in a
+new goroutine. It is risky to directly run
+
+  go supervisor.Serve()
+
+because that will briefly create a race condition as it starts up, if you
+try to .Add() services immediately afterward.
+
+*/
+type Supervisor struct {
+	Name string
+
+	spec Spec
+
+	services             map[serviceID]serviceWithName
+	cancellations        map[serviceID]context.CancelFunc
+	servicesShuttingDown map[serviceID]serviceWithName
+	lastFail             time.Time
+	failures             float64
+	restartQueue         []serviceID
+	serviceCounter       serviceID
+	control              chan supervisorMessage
+	notifyServiceDone    chan serviceID
+	resumeTimer          <-chan time.Time
+	liveness             chan struct{}
+
+	// despite the recommendation in the context package to avoid
+	// holding this in a struct, I think due to the function of suture
+	// and the way it works, I think it's OK in this case. This is the
+	// exceptional case, basically.
+	ctxMutex sync.Mutex
+	ctx      context.Context
+	// This function cancels this supervisor specifically.
+	myCancel func()
```

Comment:
Consider renaming to something more descriptive like `cancel` (this is recognizable as the context’s cancel func, due to convention) or `ctxCancel`.